protocol/op-supervisor-integration.md

```diff
@@ -0,0 +1,309 @@
+
+<flags> // 1 byte, to indicate if cross-safe was a deposit-only block
+// total: 1+8+20+4+8+20+8+20+1 = 90 bytes
+```
+
+For each chain in the dependency set, for each L1 block,
+we append one of these entries to the super-safe-db.
+If an L2 chain has not traversed gotten to the next L1 block yet, the DB writing work halts until it does.
+
+In a future iteration we may add other `<version>` data,
+and fit it into `90` bytes (using multiple entries if necessary) to register data.
+
+TODO: We modify the `event` DB to checkpoint every L2 block,
+rather than only forcefully placing such entry every 256 entries.
+This will help verify any gaps between L2 blocks,
+if the `cross-safe` or `local-safe` jumps by more than 1 block at a time.
+
+## op-supervisor RPC
+
+### `CheckMessage`
+
+```go
+type SafetyLevel string
+
+const (
+    Finalized   SafetyLevel = "finalized"
+    Safe        SafetyLevel = "safe"
+    CrossUnsafe SafetyLevel = "cross-unsafe"
+    Unsafe      SafetyLevel = "unsafe"
+	Unknown     SafetyLevel = "unknown"
+    Conflicts   SafetyLevel = "conflicts"
+)
+
+type RPC interface {
+	CheckMessage(identifier types.Identifier,
+                 payloadHash common.Hash,
+                 ) (types.SafetyLevel, error)
+}
+```
+
+This method is used to verify initiating messages exist, e.g. used by the tx-pool and block-building, to verify user-input before inclusion into the chain.
+
+API implementation:
+
+1. Find the event log-db entry at the given identifier
+2. Reconstruct a `TruncatedHash` from the `identifier` and `payloadHash`
+3. Check if the log-db entry matches the reconstructed `TruncatedHash`
+    - If it matches, return the safety-level of the entry
+        - The safety is determined by the first (and thus strongest safety level) type of head of the chain that is past or equal the entry.
+    - If it does not match, return the `conflicts` safety level.
+    - If the entry is not yet known, return an `unknown` safety level.
+
+
+### `TryCrossUnsafe`
+
+```go
+type RPC interface {
+    GetCrossUnsafe(chainID types.ChainID, maxNumber uint64) (crossUnsafe eth.BlockID, err error)
+}
+```
+
+API implementation:
+1. Compare the `maxNumber` to the locally known `unsafe` tip of the event-DB:
+    - If newer:
+      1. Initiate fetching of new event-log data.
+      2. Return the latest known `cross-unsafe`.
+    - If equal or older:
+      1. Return the latest known `cross-unsafe` block before `maxNumber`.
+         This is the last block checkpoint `<= maxNumber`
+
+By clipping using `maxNumber` the op-node is able to verify the returned `cross-unsafe`, and not accept it if it's non-canonical, and optionally query for older cross-unsafe blocks to try to bump to.
+
+### `NextDeriveTask`, `OnDerived`, `CrossSafe`
+
+```go
+type RPC interface {
+	
+	// Returns where the next L2 block has to be derived from. 
+	NextDeriveTask(chainID types.ChainID) (parentL2 eth.BlockID, depositOnly bool, l1 eth.BlockID) error
```

Comment:
To confirm: this is only about deriving local-safe blocks from L1 data. The sequencer still just uses its unsafe head to build on.

Reply:
Correct, this is just about acquiring the L1-derived -> L2 block mapping. The sequencer still builds completely independent of this derivation work.